File: codeview-ui/src/lib/server/parsing/freshness.ts
/**
 * Freshness registry — single source of truth for "have we already parsed
 * this crate at this version with this parser revision?".
 *
 * Backed by R2 keys `rust/_index/{name}.json`. One entry per crate captures
 * the version we parsed, when, with what parser, and a hash of the resulting
 * graph for idempotency. The cron-driven freshness sweep (GHA) reads these
 * entries to decide what's stale; the parse pipeline writes them on success.
 *
 * This file is the portable core — backend implementations live in
 *   `scripts/freshness-backends.ts` (Bun: miniflare SQLite + S3 SDK)
 * and (eventually) in `src/lib/server/cloudflare/freshness-binding.ts` if
 * the Worker ever needs to read freshness for UI display.
 *
 * Adapter discipline: keep the interface small (read/write/list). Anything
 * smarter (staleness predicates, batch operations) belongs in the registry
 * class, not the backend.
 */

/**
 * What we record after every successful parse. Persisted as JSON at
 * `rust/_index/{name}.json` so freshness can be inspected with a single
 * R2 GET per crate.
 *
 * `parserRevision` is the git SHA of `codeview-rustdoc` at parse time —
 * if the parser improves, the same (name, version) becomes stale and gets
 * reparsed. `schemaVersion` is the codeview graph-schema version so we can
 * force a re-parse on schema changes. `graphHash` lets the pipeline skip
 * R2 uploads when a re-parse produces byte-identical output.
 */
export interface FreshnessEntry {
	/**
	 * Canonical Rust crate name (underscored), e.g. `serde_json`. Matches the
	 * `name` used by the Rust parser and the rest of codeview-core.
	 */
	name: string;
	/**
	 * R2 storage form (hyphenated), e.g. `serde-json`. The artifact-key path
	 * is `rust/{storageName}/{version}/...` while the freshness index key is
	 * `rust/_index/{name}.json`. Capturing both here closes the loop — a UI
	 * that reads a freshness entry can directly construct the artifact URL
	 * without needing to know about the name → storageName transformation.
	 *
	 * Optional for backwards-compat: entries written before this field
	 * existed default to `name` (which matches for crates without underscores).
	 */
	storageName?: string;
	version: string;
	parsedAt: string; // ISO 8601
	source: 'docs.rs' | 'sysroot' | 'cargo' | 'unknown';
	parserRevision: string;
	schemaVersion: number;
	/**
	 * sha256 of the canonical graph JSON. Two runs with the same parser
	 * over the same input should produce the same graphHash even when
	 * the parser internally uses HashMap (insertion order varies) — the
	 * canonicaliser in `parse-one.ts` makes this stable.
	 *
	 * Used to skip R2 uploads when re-parsing produces byte-identical
	 * artifacts (parser revision changed but output didn't).
	 */
	graphHash: string;
	/**
	 * sha256 of the raw rustdoc JSON bytes from the upstream source
	 * (docs.rs gz-decompressed, or the local sysroot JSON file).
	 *
	 * Recorded for diagnostics — lets a future run notice if docs.rs
	 * served different bytes for the same crate+version (which would
	 * indicate a docs.rs rebuild against a newer toolchain). Not used
	 * as a control-flow short-circuit because every case where it
	 * could fire is already caught by the registry-level idempotency
	 * check on (version, parserRevision, schemaVersion).
	 *
	 * Optional for backwards-compat: entries written before this field
	 * existed won't have it; treat undefined as "unknown".
	 */
	rustdocHash?: string;
}

/**
 * Backend contract — anything that can read/write JSON objects under the
 * `rust/_index/` prefix qualifies. Implementations: miniflare SQLite (local
 * dev), R2 via S3 SDK (GHA + CI), R2 binding (Worker, read-only). The
 * registry never touches the backend's transport — only this interface.
 */
export interface FreshnessBackend {
	read(name: string): Promise<FreshnessEntry | null>;
	write(entry: FreshnessEntry): Promise<void>;
	list(): Promise<FreshnessEntry[]>;
}

/**
 * The R2 key shape. Exported so backends agree on the path without
 * duplicating string literals. Crate names are normalised to lowercase
 * underscored to match the rest of the artifact layout (matches
 * `normalizeCrateName` in `static-artifacts.ts`).
 */
export function freshnessKey(name: string): string {
	return `rust/_index/${name}.json`;
}

/**
 * The staleness check is its own type so callers can branch on the reason
 * (useful for cron logs: "parsing tokio because parser revision changed"
 * reads better than "parsing tokio because isStale returned true").
 */
export type StalenessReason =
	| { stale: false }
	| { stale: true; reason: 'never_parsed' }
	| { stale: true; reason: 'newer_version'; observed: string; recorded: string }
	| { stale: true; reason: 'parser_revision_changed'; recorded: string; current: string }
	| { stale: true; reason: 'schema_version_changed'; recorded: number; current: number };

/**
 * Top-level operations over the registry. Stateless wrapper around a
 * `FreshnessBackend` — the backend handles transport, the registry handles
 * predicates and book-keeping.
 *
 * Construction is just `new FreshnessRegistry(backend)`. No async setup;
 * backends are responsible for their own connection lifetimes.
 */
export class FreshnessRegistry {
	constructor(private readonly backend: FreshnessBackend) {}

	/**
	 * Decide whether a (name, observedNewestVersion) pair needs re-parsing.
	 * Stale when: never parsed | crates.io has a newer version | parser SHA
	 * differs | schema version differs.
	 *
	 * `currentParserRevision` is the git SHA the *caller* would parse with —
	 * usually `git rev-parse HEAD` on the codeview repo. Passing this in
	 * (rather than reading from disk inside) keeps the registry pure and
	 * testable without git available.
	 */
	async check(
		name: string,
		observedNewestVersion: string,
		currentParserRevision: string,
		currentSchemaVersion: number,
	): Promise<StalenessReason> {
		const entry = await this.backend.read(name);
		if (!entry) return { stale: true, reason: 'never_parsed' };
		if (entry.version !== observedNewestVersion) {
			return {
				stale: true,
				reason: 'newer_version',
				observed: observedNewestVersion,
				recorded: entry.version,
			};
		}
		if (entry.parserRevision !== currentParserRevision) {
			return {
				stale: true,
				reason: 'parser_revision_changed',
				recorded: entry.parserRevision,
				current: currentParserRevision,
			};
		}
		if (entry.schemaVersion !== currentSchemaVersion) {
			return {
				stale: true,
				reason: 'schema_version_changed',
				recorded: entry.schemaVersion,
				current: currentSchemaVersion,
			};
		}
		return { stale: false };
	}

	async record(entry: FreshnessEntry): Promise<void> {
		await this.backend.write(entry);
	}

	async latestParsed(name: string): Promise<FreshnessEntry | null> {
		return this.backend.read(name);
	}

	async listAll(): Promise<FreshnessEntry[]> {
		return this.backend.list();
	}
}

File: codeview-cli/src/publisher/freshness.rs
//! Freshness registry — single source of truth for "have we parsed
//! this crate at this version with this parser revision?".
//!
//! Backed by `rust/_index/{name}.json` keys via the [`R2`] trait, so the
//! same code path serves both local (miniflare SQLite) and remote
//! (Cloudflare R2 over S3) targets.
//!
//! Preserves the registry JSON shape: same field names, same staleness
//! predicate, same `Restricted` semantics.

use std::sync::Arc;

use anyhow::Result;
use semver::Version;
use serde::{Deserialize, Serialize};

#[cfg(test)]
use super::r2::INDEX_MANIFEST_KEY;
use super::r2::{R2, freshness_key, read_json, version_freshness_key, write_json};

/// What we record after every successful parse. JSON-on-the-wire is
/// stable: the TS UI hits these files for "last parsed at" decorations.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FreshnessEntry {
    pub name: String,
    /// R2 storage form (hyphenated), e.g. `serde-json`. Set when the
    /// canonical Rust name differs from the on-disk path so consumers
    /// can construct artifact URLs without re-running the
    /// `normalise → hyphenate` transform.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub storage_name: Option<String>,
    pub version: String,
    /// ISO 8601 timestamp.
    pub parsed_at: String,
    pub source: Source,
    /// Git SHA of the parser at parse time. Bumps when `codeview-rustdoc`
    /// changes, marking every entry stale and triggering a re-parse on
    /// the next cron sweep.
    pub parser_revision: String,
    /// Graph-schema version (mirrors `codeview-core::SCHEMA_VERSION`).
    pub schema_version: u32,
    /// sha256 of the canonical graph JSON. Lets the orchestrator skip
    /// R2 uploads when re-parsing produces byte-identical artifacts
    /// (parser revision changed but output didn't).
    pub graph_hash: String,
    /// sha256 of the raw rustdoc JSON bytes from docs.rs. Recorded for
    /// drift detection ("did docs.rs serve different bytes for the
    /// same version after a toolchain bump?").
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rustdoc_hash: Option<String>,
    pub nodes: usize,
    pub edges: usize,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum Source {
    #[serde(rename = "docs.rs")]
    DocsRs,
    Sysroot,
    Cargo,
    Unknown,
}

/// Why the registry says a crate needs re-parsing — carries the previous
/// and current values so callers can log them.
#[derive(Debug, Clone)]
pub enum Staleness {
    Fresh,
    NeverParsed,
    NewerVersion { observed: String, recorded: String },
    ParserRevisionChanged { recorded: String, current: String },
    SchemaVersionChanged { recorded: u32, current: u32 },
}

impl Staleness {
    pub fn is_stale(&self) -> bool {
        !matches!(self, Staleness::Fresh)
    }

    pub fn describe(&self) -> String {
        match self {
            Staleness::Fresh => "fresh".into(),
            Staleness::NeverParsed => "never parsed".into(),
            Staleness::NewerVersion { observed, recorded } => {
                format!("crates.io {observed} vs recorded {recorded}")
            }
            Staleness::ParserRevisionChanged { recorded, current } => {
                format!(
                    "parser {} → {}",
                    &recorded[..recorded.len().min(8)],
                    &current[..current.len().min(8)],
                )
            }
            Staleness::SchemaVersionChanged { recorded, current } => {
                format!("schema v{recorded} → v{current}")
            }
        }
    }
}

/// Top-level operations over the registry. Stateless wrapper around an
/// `Arc<dyn R2>` — the trait handles transport, this type handles
/// staleness predicates and key shaping.
pub struct FreshnessRegistry {
    r2: Arc<dyn R2>,
}

impl FreshnessRegistry {
    pub fn new(r2: Arc<dyn R2>) -> Self {
        Self { r2 }
    }

    /// Read one entry by canonical crate name.  `Ok(None)` if the key
    /// is absent — distinguished from transport errors.
    pub async fn latest(&self, name: &str) -> Result<Option<FreshnessEntry>> {
        read_json(&self.r2, &freshness_key(name)).await
    }

    /// Read one exact crate version freshness entry.
    pub async fn version(&self, name: &str, version: &str) -> Result<Option<FreshnessEntry>> {
        read_json(&self.r2, &version_freshness_key(name, version)).await
    }

    /// Decide whether `(name, observed_newest_version)` needs
    /// re-parsing. Stale on first miss, newer version on crates.io,
    /// parser-SHA bump, or schema bump.
    pub async fn check(
        &self,
        name: &str,
        observed_newest_version: &str,
        current_parser_revision: &str,
        current_schema_version: u32,
    ) -> Result<Staleness> {
        let Some(entry) = self.version(name, observed_newest_version).await? else {
            return Ok(Staleness::NeverParsed);
        };
        if entry.version != observed_newest_version {
            return Ok(Staleness::NewerVersion {
                observed: observed_newest_version.to_string(),
                recorded: entry.version,
            });
        }
        if entry.parser_revision != current_parser_revision {
            return Ok(Staleness::ParserRevisionChanged {
                recorded: entry.parser_revision,
                current: current_parser_revision.to_string(),
            });
        }
        if entry.schema_version != current_schema_version {
            return Ok(Staleness::SchemaVersionChanged {
                recorded: entry.schema_version,
                current: current_schema_version,
            });
        }
        Ok(Staleness::Fresh)
    }

    pub async fn record(&self, entry: &FreshnessEntry) -> Result<()> {
        write_json(
            &self.r2,
            &version_freshness_key(&entry.name, &entry.version),
            entry,
        )
        .await?;

        let latest = self.latest(&entry.name).await?;
        if should_replace_latest(latest.as_ref(), entry) {
            write_json(&self.r2, &freshness_key(&entry.name), entry).await?;
        }
        Ok(())
    }

    /// Enumerate every recorded crate.  Used by the catalog rebuilder.
    pub async fn list_all(&self) -> Result<Vec<FreshnessEntry>> {
        let mut out = Vec::new();
        let keys = self.r2.list_prefix("rust/_index/by-version/").await?;
        for key in keys {
            if !is_version_freshness_key(&key) {
                continue;
            }
            if let Some(entry) = read_json::<FreshnessEntry>(&self.r2, &key).await? {
                out.push(entry);
            }
        }
        Ok(out)
    }
}

fn should_replace_latest(current: Option<&FreshnessEntry>, candidate: &FreshnessEntry) -> bool {
    let Some(current) = current else {
        return true;
    };
    compare_versions_desc(&candidate.version, &current.version).is_le()
}

fn compare_versions_desc(left: &str, right: &str) -> std::cmp::Ordering {
    match (Version::parse(left), Version::parse(right)) {
        (Ok(left), Ok(right)) => right.cmp(&left),
        (Ok(_), Err(_)) => std::cmp::Ordering::Less,
        (Err(_), Ok(_)) => std::cmp::Ordering::Greater,
        (Err(_), Err(_)) => right.cmp(left),
    }
}

#[cfg(test)]
fn is_per_crate_freshness_key(key: &str) -> bool {
    key.starts_with("rust/_index/")
        && key.ends_with(".json")
        && key != INDEX_MANIFEST_KEY
        && !key.starts_with("rust/_index/_generations/")
        && !key.starts_with("rust/_index/by-version/")
}

fn is_version_freshness_key(key: &str) -> bool {
    key.starts_with("rust/_index/by-version/") && key.ends_with(".json")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn manifest_crate_freshness_key_does_not_collide_with_aggregate_manifest() {
        assert_eq!(freshness_key("manifest"), "rust/_index/manifest.json");
        assert!(is_per_crate_freshness_key(&freshness_key("manifest")));
        assert!(!is_per_crate_freshness_key(INDEX_MANIFEST_KEY));
        assert!(!is_per_crate_freshness_key(
            "rust/_index/_generations/gen/shards/00.json"
        ));
        assert!(!is_per_crate_freshness_key(
            "rust/_index/by-version/serde/1.0.0.json"
        ));
        assert!(is_version_freshness_key(
            "rust/_index/by-version/serde/1.0.0.json"
        ));
    }
}
